app/src/main/java/com/amaze/filemanager/filesystem/operations/AbstractOperation.java

```diff
@@ -0,0 +1,67 @@
+package com.amaze.filemanager.filesystem.operations;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+import androidx.annotation.CallSuper;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.WorkerThread;
+
+/**
+ * Any task that needs an undo can be encapsulated with an AbstractOperation,
+ *  this, with the Operator, allows for easy sequencing of recursive operations,
+ *  by just calling requires()
+ */
+public abstract class AbstractOperation {
+	private Operator operator;
+
+	/**
+	 * To be used only by the Operator for dependency injection
+	 */
+	/* package-protected */ void setOperator(@NonNull Operator operator) {
+		this.operator = operator;
+	}
+
+	private Operator getOperator() {
+		return operator;
+	}
+
+	/**
+	 * If this check passes the operation will be started,
+	 * nothing here is reversable and this operation can have no side effects!
+	 */
+	protected abstract boolean check();
+
+	/**
+	 * The proper operation,
+	 * if another operation is required to be started call requires() and it will eventually start
+	 */
```

Comment:
"When this function is called, all prerequisite operations will execute 
before this operation executes." that is not correct, requires runs code _after the operation ends_.

Reply:
So if a caller calls `op1.requires(op2)`, what is the correct order here?
```
a) op1 -> op2
b) op2 -> op1
```

My understanding is that a prerequisite task is checked by calling `op.check()` to ensure the Operation is ready to be executed, but I thought `requires` adds prerequisite tasks?

Reply:
> So if a caller calls `op1.requires(op2)`, what is the correct order here?
> 
> ```
> a) op1 -> op2
> b) op2 -> op1
> ```

a

> My understanding is that a prerequisite task is checked by calling `op.check()` to ensure the Operation is ready to be executed, and I thought `requires` adds prerequisite tasks?

I made mistakes choosing the correct words. 

`check` makes sure the current state of the filesystem is correct for the operation that is going to be performed, and check that the operation can actually be performed.
`requires` ads post-requisite tasks.

This is done this way so that operations of folders (that are normally recursive) can be done sequentially.

For example: copying:
	_Check_ stuff, i.e. that the destination is writable.
	If copying a folder: Create parent folder. -> _require_ copy of children.
	If copying a file: Create child, copy contents.

If the `require` wasn't there the operation would be recursive, as is exists it can be run sequentially by the Operator.